packages/h5wasm/src/utils.ts

```diff
@@ -19,73 +20,66 @@ export function isHDF5(buffer: ArrayBuffer): boolean {
   );
 }
 
-export function isH5WasmGroup(entity: H5WasmEntity): entity is H5WasmGroup {
-  return entity instanceof H5WasmGroup;
-}
-
-export function isH5WasmDataset(entity: H5WasmEntity): entity is H5WasmDataset {
-  return entity instanceof H5WasmDataset;
-}
-
-export function isH5WasmSoftLink(
-  entity: H5WasmEntity
-): entity is H5WasmSoftLink {
-  return entity instanceof H5WasmSoftLink;
-}
+export function convertMetadataToDType(metadata: NumericMetadata): NumericType;
+export function convertMetadataToDType(metadata: Metadata): DType;
+export function convertMetadataToDType(metadata: Metadata): DType {
+  if (isIntegerMetadata(metadata)) {
```

Comment:
I reworked the switch to use dedicated guards to allow TS to infer the typing of metadata.

It allows more specific destructurings for Integer, Float and String but the important part is that compound metadata is correctly inferred.